api/maven-api-core/src/main/java/org/apache/maven/api/PathType.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.api;
+
+import java.nio.file.Path;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.apache.maven.api.annotations.Experimental;
+import org.apache.maven.api.annotations.Nonnull;
+
+/**
+ * The option of a command-line tool where to place the paths to some dependencies.
+ * A {@code PathType} can identify the Java class-path, the Java module-path,
+ * or another kind of path for another programming language for example.
+ * This class is like an enumeration, except that it is extensible:
+ * plugins can define their own kinds of path.
+ *
+ * <p>Path types are often exclusive. For example, a dependency should not be both
+ * on the Java class-path and on the Java module-path.</p>
+ *
+ * @see DependencyProperties#PATH_TYPES
+ * @see org.apache.maven.api.services.DependencyResolverResult#getDispatchedPaths()
+ *
+ * @since 4.0.0
+ */
+@Experimental
+public abstract class PathType {
```

Comment:
I'm not convinced we will ever have an implementation other than `JavaPathType`, in which case, both interfaces could be merged.  I'd also prefer an interface rather than an abstract class if that's possible.

Reply:
We can take an approach similar to `java.nio.file`, where `OpenOption` is an interface and `StandardOpenOption` is an `enum` implementing that interface. So no problem for making `PathType` an interface.

`JavaPathType` (or `PathType` if merged) would have been an `Enum` if it wasn't for the need to be extensible. Enumerations have to be classes, otherwise we cannot define the enumeration values in the API.

Reply:
Makes `PathType` an interface, `JavaPathType` an enumeration implementing that interface, and the case of the `--patch-module` option an inner class named `Modular` also implementing the `PathType` interface. The latter class generalizes `--path-module` support to other kinds of module-dependent option if some appear in the future.

Reply:
Not sure we need the enumeration at all, java is not in a closed set of paths (just check out war, ear, osgi for common examples but this is still just a subset) so in this design type is really a banalised value core does not even have to be aware about at the end of the refactoring, only mojos (even if shared somewhere but not core) so an interface is okish even if it is really just a `String` and not name+option.
The link to an option is dependent of the context (=mojo) so would rather belong to the impl (integration layer since we wouldnt have an interface for a String I guess) - this is one reason I was not ultra enthusiast to use type for that, means we need to ensure we can duplicate dependencies with multiple types now even if all are just `jar`.

From a design standpoint this is too opiniated to stay in core IMHO as mentionned on the list so maybe keep it an `interface {name}` or merge whatever works then work on moving it outside core (not 100% sure maven-shared is still a thing for plugins but looks like it would be a saner location or we just create a repo for java plain command line options maybe).

Reply:
The `JavaPathType` focuses on the standard Java tools. In order to avoid being opinionated, the current strategy is to take the [Java Development Kit Version 21 Tool Specifications](https://docs.oracle.com/en/java/javase/21/docs/specs/man/index.html) and put all options for all kinds of path in the `JavaPathType` enumeration. So this is non-arbitrary from a Java standard point of view.

If there is a demand for other popular types of path, I suppose that they could be added to this enumeration as well. But even if they are not added, all plugins have the possibility to provide their own `PathType` enumeration if they want.

The `PathType.option()` method could indeed be controversial. The reason for this API is because some options are formatted is a special way. Currently, `--patch-module` is the only example I'm aware of, but more may exist. This architecture makes possible to handle those special cases in dedicated `PathType` implementations. The presence of this API in a shared place is because _Java Development Kit Version 21 Tool Specifications_ uses those options consistently in all tools as far as I know. For `PathType` defined in plugins, the plugin has the possibility to provide the right option for the tool that they target.

Note: `JavaPathType` is currently not the only Java-specific interface in core. There is also `JavaToolchain`.

Reply:
> So this is non-arbitrary from a Java standard point of view.

well, with examples I gave it is, you target a few java standalone tools - agree it is the most used in builds but it is not a reason IMHO and stays bad to link core to spceific plugins so much (we already suffer from that).

> I suppose that they could be added to this enumeration as well

So what you propose is to include compiler, javadoc, jlink, jpackage etc into maven core and drop plugins (I know it is not what you wrote but it is what it means otherwise the enum is not correct as explained before).

> The PathType.option() method could indeed be controversial.

It is well understood but here again linked to specific tools to in maven it belongs to plugins or their transitive deps ("shared" module), not to core which is just a plugin orchestrator from a design standpoint, less we have there better we'll be IMHO.

Reply:
> well, with examples I gave it is

No, Java is a registered trademark protected by laws. Only Oracle controls what can be call "Java® specification". Everything else is non-standard, unless agreed by Oracle (e.g. through JSP). It may be widely-used, but it is not standard.

> what you propose is to include compiler, javadoc, jlink, jpackage etc into maven core and drop plugins 

No. There is a plugin for `javac`, another plugin for `javadoc`, _etc._ But all those plugins share the notion of class-path, module-path, _etc._ The path type is a cross-plugins concept, which is why it needs to be defined in a common place. That's all, only the common stuff is defined there.

Reply:
Well you go nowhere there since some of the brands i mzntionned were owned by the same guys and ultimately if you are reasoning like that maven is no limited to java so same state for me.

The cross plugin need was ack, no need to discuss it and you got the reason why it doesnt belong to core with an example that you would pull all the code of all plugins in core, even osgi for the same reason.
We just need it in a lib and find a neat way to change type/option mapping per plugin as explained on the list - same module.

Reply:
There is an official thing called _Java® tools specification_. Its content is the official standard according those who legally own the "Java" name for the programming language. Everything else is not Java. It may be called "OpenJDK", "Eclipse" or "Adoptium", but not Java. Whether the extra stuff is done by the same peoples or not is irrelevant to the _Java® tools specification_.

The _Java® tools specification_ is only taken as a base. Any library or plugin can define its own `PathType`. The _Java® tools specification_ is considered a base worth to be in shared code because some of its concepts, such as class-path, are shared by almost all tools. This is not the case of WAR, EAR, OSGi, etc.

We are not going to move plugins in core. We put there only the parts that are common to almost all Java tools, which is path management now, maybe `--add-exports` management later. Everything else stay in their plugins.

If you want Maven core to be language neutral and move all Java-specific stuff in a separated module shared by all Java plugins, I have no objection. But this is a separated debate (`JavaPathType` is not the only Java-specific thing currently in core).

Reply:
> There is an official thing called Java® tools specification

True but this is unrelated to this PR by the fact maven does not aim to target this scope but way more so question is do we put all we intend to support in core or not.
For me answer can only be no otherwise core would quickly be bloated and we already saw it does not scale in time (in terms of maintenance).
So overall you can reference any trademark but this stays out of topic for me, topic is what is the scope of core and it is the pipeline only for me and hopefully we'll be able to drop mvn 2 legacy remaining parts not respecting that historically.

> This is not the case of WAR, EAR, OSGi, etc.

Once again it is out of topic but there are a lot of shared parts there too so if you reasoning is right then we must embrace that too (it is default maven plugins too after all).

> If you want Maven core to be language neutral and move all Java-specific stuff in a separated module shared by all Java plugins, I have no objection.

Exactly.

> But this is a separated debate (JavaPathType is not the only Java-specific thing currently in core).

Not really, the point is that `JavaPathType` has not to exist in core but in `maven-java-tool-shared` module (whatever name it is). Core can stick to `String` since `option` doesnt always makes sense.

Reply:
It has never been proposed to put everything in core, but **only the base shared by practically all Java tools**. This base is defined by the _Java® tools specification_ (or a subset of it), not OSGi, not WAR, not EAR. If you want to add some extra paths shared by the latter, it can be discussed. But as starting point, the _Java® tools specification_ provides a standard definition of what a base is.

I have no objection in moving Java-specific stuff in a separated module. But it cannot be done as long as methods such as `isAddedToClasspath()` are in core, even if they are legacy (otherwise we cannot implement them as a redirection to new API). If those methods move, I will follow. But my impression is that Maven has close ties to Java, both in its API and inherited from `org.eclipse.aether` API.

Reply:
> Not really, the point is that JavaPathType has not to exist in core but in maven-java-tool-shared module (whatever name it is). Core can stick to String since option doesnt always make sense.

Core _always had the "stuff" to describe WHAT you use (Artifact, ArtifactHandler)_. Plugins in turn _have the stuff/logic that knows HOW to use them_. This PR is to extend the vocabulary for "what" part. Your reasoning to me sounds like even Artifact should not be in core, and every plugin should define for itself what "artifact" is.

Java was always part of the core, so do not come now with some "shared java lang library" please :smile:  THAT is totally and utterly off topic for this PR (to make Maven fully and totally lang agnostic).

Reply:
@cstamas it is what it is, most of maven stuff related to java is inherited from a time maven was more bloated but I don't find a single point to pursue that path, even now, in particular cause each plugin needs its own configuration now (it was explained on the list) so more we put in core more we'll bring whatever the original will.

> only the base shared by practically all Java tools

Does not mean much sadly and is either a too small subset or a too big subset but here we don't reach any goal with something in between.

Just let move it to a shared module and move from there.

On my side I'm more than fine dropping `isAddedToClasspath` is maven 4 API, it does not make much sense with your PR (:tada:).

> THAT is totally and utterly off topic for this PR

fine with that while it does not end in any release, can be done after while only snapshots for me